core/src/main/java/org/apache/calcite/rel/rules/MinusToFilterRule.java

```diff
@@ -0,0 +1,133 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Minus;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.Pair;
+
+import com.google.common.collect.ImmutableList;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.immutables.value.Value;
+
+/**
+ * Rule that replaces {@link Minus} operator with {@link Filter}
+ * when both inputs are from the same source with only filter conditions differing.
+ *
+ * <p>Example transformation:
+ * <blockquote><pre>
+ * SELECT mgr, comm FROM emp WHERE mgr = 12
+ * EXCEPT
+ * SELECT mgr, comm FROM emp WHERE comm = 5
+ *
+ * to
+ *
+ * SELECT DISTINCT mgr, comm FROM emp
+ * WHERE mgr = 12 AND NOT(comm = 5)
+ * </pre></blockquote>
+ */
+@Value.Enclosing
+public class MinusToFilterRule
+    extends RelRule<MinusToFilterRule.Config>
+    implements TransformationRule {
+
+  /** Creates an MinusToFilterRule. */
+  protected MinusToFilterRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Minus minus = call.rel(0);
+    if (minus.all || minus.getInputs().size() != 2) {
+      return;
+    }
+    final RelBuilder builder = call.builder();
+    final RelNode leftInput = call.rel(1);
+    final Filter rightInput = call.rel(2);
+
+    if (!RexUtil.isDeterministic(rightInput.getCondition())) {
+      return;
+    }
+
+    final Pair<RelNode, @Nullable RexNode> leftBase = getUnfilteredInputAndMergeCond(leftInput);
+    final Pair<RelNode, @Nullable RexNode> rightBase = getUnfilteredInputAndMergeCond(rightInput);
+    if (!leftBase.left.equals(rightBase.left)) {
+      return;
+    }
+
+    // Right input is Filter, right cond should be not null
+    assert rightBase.right != null;
+    final RexNode finalCond = leftBase.right != null
+        ? builder.and(leftBase.right, builder.not(rightBase.right))
+        : builder.not(rightBase.right);
+
+    builder.push(leftBase.left)
+        .filter(finalCond)
+        .distinct();
+
+    call.transformTo(builder.build());
+  }
+
+  private static Pair<RelNode, @Nullable RexNode> getUnfilteredInputAndMergeCond(
+      final RelNode node) {
+    RexBuilder builder = node.getCluster().getRexBuilder();
+    RelNode stripped = node.stripped();
+    RexNode cond = null;
+    while (stripped instanceof Filter) {
```

Comment:
here you handle nested filters; perhaps you should document this in the JavaDoc.

Reply:
Done!

Reply:
I mean the JavaDoc for the rule.

Reply:
Done!